models/models/markov_chain/markov_chain.py

```diff
@@ -42,7 +44,15 @@ def train(self, epochs: int, xtrain: Any, ytrain: Any, progress_callback: Progre
         for i in range(n_gram_next.shape[0]):
             self.probabilities[i] = {}
 
+
+        start = time.time()
+        time.perf_counter()
+
         for i in range(len(n_gram_next)):
+
+            elapsed = time.time() - start
+            progress_callback(SeriesProgress([(elapsed, 100*i/len(n_gram_next))]))
```

Comment:
FYI you dont have to specify `SeriesProgress`, feel free to ignore this comment

```suggestion
            progress_callback([(elapsed, 100*i/len(n_gram_next))])
```